src/main/java/org/terasology/metalrenegades/interaction/systems/WellWaterSystem.java

```diff
@@ -15,36 +15,146 @@
+                    wellSourceComp.waterRefills = wellSourceComp.maxRefills;
+                }
+                waterSource.send(new WellRefilledEvent());
+            }
+            worldTimeCycles = 0;
+        }
+
+        worldTimeCycles++;
+    }
+
+    @ReceiveEvent(components = {WellBlockComponent.class})
+    public void onActivate(ActivateEvent event, EntityRef sourceBlock) {
         EntityRef gatheringCharacter = event.getInstigator();
-        EntityRef cupItem = entityManager.create("MetalRenegades:waterCup");
+        EntityRef heldItem = gatheringCharacter.getComponent(CharacterHeldItemComponent.class).selectedItem;
+        if (!gatheringCharacter.exists()) {
+            return;
+        }
+
+        EntityRef wellEntity = getWellEntity(sourceBlock);
+        if (wellEntity == null) {
+            logger.warn("No well entity found for activated well block!");
+            return;
+        }
+
+        WellSourceComponent wellSourceComp = wellEntity.getComponent(WellSourceComponent.class);
+        if (wellSourceComp.waterRefills <= 0) { // if no refills remain, don't give water.
+            return;
+        }
+        wellSourceComp.waterRefills--;
+        wellEntity.saveComponent(wellSourceComp);
+
+        if (!heldItem.hasComponent(WaterCupComponent.class)) {
+            ThirstComponent thirst = gatheringCharacter.getComponent(ThirstComponent.class);
+
+            thirst.lastCalculatedWater = thirst.maxWaterCapacity;
+            thirst.lastCalculationTime = time.getGameTimeInMs();
+            gatheringCharacter.saveComponent(thirst);
+
+            wellEntity.send(new WellDrinkEvent(gatheringCharacter));
 
-        if(!cupItem.exists() || !gatheringCharacter.exists()) {
             return;
         }
 
+        EntityRef cupItem = entityManager.create("MetalRenegades:waterCup");
         cupItem.send(new GiveItemEvent(gatheringCharacter));
+        wellEntity.send(new CupFilledEvent(gatheringCharacter, cupItem));
+        heldItem.destroy();
+    }
+
+    @ReceiveEvent(components = {WaterCupComponent.class})
+    public void onDrinkConsumed(DrinkConsumedEvent event, EntityRef item) {
+        EntityRef drinkingCharacter = event.getInstigator();
+        EntityRef cupItem = entityManager.create("MetalRenegades:emptyCup");
+
+        if (!cupItem.exists() || !drinkingCharacter.exists()) {
+            return;
+        }
+
+        item.destroy();
+        cupItem.send(new GiveItemEvent(drinkingCharacter));
+    }
+
+    /**
+     * Retrieves the well entity associated with a well water source block.
+     *
+     * @param sourceBlock The activated source block.
+     * @return The well entity that contains this block.
+     */
+    private EntityRef getWellEntity(EntityRef sourceBlock) {
+        LocationComponent blockLocComp = sourceBlock.getComponent(LocationComponent.class);
+        Vector3i blockLocation = new Vector3i(blockLocComp.getWorldPosition());
+        EntityRef wellEntity = null;
```

Comment:
This looks wrong - I would think it would be better to use `EntityRef.NULL` here.

Reply:
Agreed, this has been changed to the EntityRef version 👍